tests/PHPUnit/Integration/Segment/SegmentsCacheTest.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Tests\Integration;
+
+use Piwik\Cache;
+use Piwik\Segment;
+use Piwik\Tests\Framework\TestCase\IntegrationTestCase;
+use Piwik\Plugins\SegmentEditor\API as SegmentEditorApi;
+use Piwik\Plugins\SegmentEditor\Model as SegmentEditorModel;
+
+/**
+ * @group Segment
+ * @group SegmentsCacheTest
+ */
+class SegmentsCacheTest extends IntegrationTestCase
+{
+    public function test_segment_hash_cached()
+    {
+        $cache = Cache::getEagerCache();
+        $model = new SegmentEditorModel();
+
+        $idSegment = SegmentEditorApi::getInstance()->add('test segment 1', 'browserCode==ff');
+        $segment = $model->getSegment($idSegment);
+        $key1 = Segment::CACHE_KEY . md5($segment['definition']);
+
+        $this->assertFalse($cache->contains($key1));
+
+        Segment::getSegmentHash('dummy');
+
+        $this->assertTrue($cache->contains($key1));
+    }
+
+    public function test_if_segment_not_found_default_hash_returned()
+    {
+        $definition = 'browserCode==ch';
+        $defaultHash = md5(urldecode($definition));
+
+        $this->assertEquals($defaultHash, Segment::getSegmentHash($definition));
+    }
+
+    public function test_cashed_hashes_for_similar_segments()
+    {
+        $cache = Cache::getEagerCache();
+        $model = new SegmentEditorModel();
+
+        // pageUrl==http://abc.com/d+f
+        // pageUrl==http://abc.com/d f
+        // pageUrl==http://abc.com/d%20f
+        $definitions = [
+            'pageUrl==http%253A%252F%252Fabc.com%252Fd%252Bf',
+            'pageUrl==http%253A%252F%252Fabc.com%252Fd%2520f',
+            'pageUrl==http%253A%252F%252Fabc.com%252Fd%252520f',
+        ];
+
+        $idSegment1 = SegmentEditorApi::getInstance()->add('test segment 1', $definitions[0]);
+        $idSegment2 = SegmentEditorApi::getInstance()->add('test segment 2', $definitions[1]);
+        $idSegment3 = SegmentEditorApi::getInstance()->add('test segment 3', $definitions[2]);
+        $segment1 = $model->getSegment($idSegment1);
+        $segment2 = $model->getSegment($idSegment2);
+        $segment3 = $model->getSegment($idSegment3);
+        $key1 = Segment::CACHE_KEY . md5($segment1['definition']);
+        $key2 = Segment::CACHE_KEY . md5($segment2['definition']);
+        $key3 = Segment::CACHE_KEY . md5($segment3['definition']);
+
+        Segment::getSegmentHash('dummy');
+
+        $this->assertEquals($segment1['hash'], $cache->fetch($key1));
+        $this->assertEquals($segment2['hash'], $cache->fetch($key2));
+        $this->assertEquals($segment3['hash'], $cache->fetch($key3));
```

Comment:
Can you also add assertions to check the contents of these cache keys (for each segment):

```
            $cacheKeyTemp = self::CACHE_KEY . md5(urldecode($segment['definition']));
            $cache->save($cacheKeyTemp, $segment['hash']);

            $cacheKeyTemp = self::CACHE_KEY . md5(urlencode($segment['definition']));
            $cache->save($cacheKeyTemp, $segment['hash']);
```
?

Reply:
I updated the tests, still green 😄  I wasn't sure TBH, but that's why we need tests 🚀